File: app/routes/services/shopify/fetchProducts.ts
import { json } from "@remix-run/node";
import { ShopifyProduct } from "../../types";

export const fetchShopifyProducts = async (shopifyStore: string, shopifyAccessToken: string): Promise<ShopifyProduct[]> => {
  const query = `
    query {
      products(first: 10) {
        edges {
          node {
            id
            title
            description
            handle
            onlineStoreUrl
            variants(first: 1) {
              edges {
                node {
                  price
                }
              }
            }
            images(first: 1) {
              edges {
                node {
                  src
                }
              }
            }  
          }
        }
      }
    }
  `;

  try {
    const response = await fetch(`https://${shopifyStore}/admin/api/2023-10/graphql.json`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Shopify-Access-Token": shopifyAccessToken,
      },
      body: JSON.stringify({ query }),
    });

    if (!response.ok) {
      return json({ error: "Failed to fetch products" }, { status: 500 });
    }

    const data = await response.json();
    return data.data.products.edges.map((edge: any) => ({
      id: edge.node.id,
      title: edge.node.title,
      description: edge.node.description,
      url: edge.node.url || `https://${shopifyStore}/products/${edge.node.handle}`,
      price: edge.node.variants.edges[0]?.node.price,
      image: edge.node.images.edges[0]?.node.src,
    }));
  } catch(error) {
    console.error(error);
    throw new Error("Failed to fetch products from Shopify");
  }
};

File: api/services/shopify/fetch_products.py
import httpx
from typing import List
from schemas.models import ShopifyProduct

async def fetch_shopify_products(shopify_store: str, shopify_access_token: str) -> List[ShopifyProduct]:
    query = """
      query {
        products(first: 10) {
          edges {
            node {
              id
              title
              description
              category {
                name
              }
              handle
              onlineStorePreviewUrl
              variants(first: 1) {
                edges {
                  node {
                    price
                  }
                }
              }
              media(first: 1) {
                edges {
                  node {
                    id
                    preview {
                      image {
                        url
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    """

    try:
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"https://{shopify_store}/admin/api/2023-10/graphql.json",
                headers={
                    "Content-Type": "application/json",
                    "X-Shopify-Access-Token": shopify_access_token,
                },
                json={"query": query}
            )

            response.raise_for_status()
            data = response.json()

            products = []
            for edge in data["data"]["products"]["edges"]:
                node = edge["node"]

                image_url = ""
                if node["media"]["edges"]:
                    image_url = node["media"]["edges"][0]["node"]["preview"]["image"]["url"]

                products.append(ShopifyProduct(
                    id=node["id"],
                    title=node["title"],
                    description=node.get("description", ""),
                    category=node["category"]["name"] if node.get("category") else None,
                    url=node.get("onlineStorePreviewUrl"),
                    price=node["variants"]["edges"][0]["node"]["price"],
                    image=image_url
                ))

            return products
    except Exception as error:
        print(f"Error fetching products: {error}")
        raise ValueError("Failed to fetch products from Shopify")